### ISSUE TITLE
Adam does not find unique global minimum with JAX 0.7.1 + double precision if PyTree not 1D array

### ISSUE THREAD

**Issue by johannahaffner** [google-deepmind/optax#1412]
Hi,

and thanks for the amazing library! 

It appears that some update in JAX 0.7.1 causes a substantial pessimisation for standard Adam on a toy problem, which fails to converge to the unique global minimum of the Rosenbrock function within 10k iterations. The shape of the PyTree matters: I get the expected result if I initialise with `1.5 * jnp.ones((20, 1))` instead of with `1.5 * jnp.ones((10, 2))`. 

I have reduced this to the MWE below

```python
import jax
import jax.numpy as jnp
import optax


jax.config.update("jax_enable_x64", True)


def rosenbrock(y):
    y_flat = y.flatten()
    residuals = (100 * (y_flat[1:] - y_flat[:-1]), 1.0 - y_flat[:-1])
    return 0.5 * jnp.sum(jnp.asarray(residuals)**2)


def _loop(init):
    optimizer = optax.adam(learning_rate=3e-3)
    opt_state = optimizer.init(init)

    def cond_fun(carry):
        _, step, _, _ = carry
        return step < 10_000

    def body_fun(carry):
        y, step, opt_state, prev_loss = carry
        loss, grads = jax.value_and_grad(rosenbrock)(y)
        # jax.debug.print("grads: {}", grads)  # Uncommenting fixes it
        updates, new_opt_state = optimizer.update(grads, opt_state, y)
        new_y = optax.apply_updates(y, updates)
        return (new_y, step + 1, new_opt_state, loss)

    init_state = (init, 0, opt_state, jnp.inf)
    final_state = jax.lax.while_loop(cond_fun, body_fun, init_state)
    final_y, final_steps, _, final_loss = final_state

    return final_y, final_steps, final_loss


init = 1.5 * jnp.ones((10, 2))

final_y, steps, final_loss = _loop(init)
print(f"Final loss: {final_loss:.8f}, expected: ~0.0")
# 0.08863456 with JAX 0.7.1, 0.00000003 with 0.7.0
```

This issue first came up as a failure in our test suite in https://github.com/patrick-kidger/optimistix/issues/160, but it does not depend on Optimistix or Equinox. It may not depend on optax either, given that forcing materialisation with a `jax.debug` statement causes the issue to go away. 

Do you have any insight as to how this behaviour might come about?

**Comment by rdyro** [comment]
After bisecting commits in JAX, this behavior seems to maybe come from changes in jjaxlib.

I really appreciate the repro. It's not clear to me how much numerical stability we expect from the rosenbrock function optimization.  It's rather surprising float64 does not reach a global minimum, but float32 does for all jaxlib versions I tested.

Btw, a faster way to ensure gradients are realized is probably
```python
# jax.debug.print("grads: {}", grads)  # Uncommenting fixes 
grads = jax.lax.optimization_barrier(grads)
```
it also gives
```
Final loss: 0.00000003, expected: ~0.0
```

Could you open an issue in JAX? I can do it otherwise.

**Comment by johannahaffner** [comment]
> After bisecting commits in JAX, this behavior seems to maybe come from changes in jjaxlib.

Thank you for bisecting this! Could you narrow down the time in which this change may have occurred?

>  It's rather surprising float64 does not reach a global minimum, but float32 does for all jaxlib versions I tested.

Rosenbrock is indeed numerically challenging, but I would expect an improved (if more expensive) solution at higher precision. For JAX 0.7.0, I get the same final loss as with a `jax.debug.print`. 

> Could you open an issue in JAX? I can do it otherwise.

Very happy to! I'll include your suggestion to use the optimisation barrier, I did not know this was a thing.

**Comment by johannahaffner** [comment]
PS: I'm going to give a pure JAX MWE a shot, currently trying to get optax specifics out of the repro :)

**Comment by rdyro** [comment]
> Thank you for bisecting this! Could you narrow down the time in which this change may have occurred?

I'll take a look soon internally.

> Rosenbrock is indeed numerically challenging, but I would expect an improved (if more expensive) solution at higher precision. For JAX 0.7.0, I get the same final loss as with a jax.debug.print.

Definitely, this is definitely not expected!

> PS: I'm going to give a pure JAX MWE a shot, currently trying to get optax specifics out of the repro :)

I appreciate that, but your optax repro is already concise, so that should be enough, thanks!

**Comment by johannahaffner** [comment]
> I'll take a look soon internally.

Awesome! 

> I appreciate that, but your optax repro is already concise, so that should be enough, thanks!

...and I did not see that! Haha I could have done something else :D hope its helpful though.

**Comment by rdyro** [comment]
Some additional discussion here: https://github.com/jax-ml/jax/issues/31473
